Improve clarity of `animation-composition` (#44431)
mdn/content: files/en-us/web/css/reference/properties/animation-composition/index.md

Reviewer: ```suggestion
Here the underlying value is `translateX(30px) rotate(45deg)`.
```

Reviewer: You should also explain what the effect value is at each stage

Reviewer: One issue I have with this demo is that the layout is broken in Firefox (the third container falls down to its own line). You could fix this by removing the fixed `width` on the `.container` elements, instead setting `flex: 1` on them, and then setting `display: flex; gap: 3px` on the `body`.

Author: The 3 containers are all in a single line on Firefox for me at a normal 16:9 1080p viewport. When narrowed, the `accumulate` naturally goes to its own line on all browsers for me, not just Firefox. It's scrollable though - is this still an issue we need to address? Could add a `margin-block` or make it a Grid just so there's block spacing when `accumulate` goes to its own line.

The issue with making them flex items is that the animated elements will overflow at a couple of viewport width ranges, including mobile.

Reviewer: @mao-sz good point. OK, I've found a better solution that works on mobile:

```css
body {
  display: flex;
  flex-wrap: wrap;
  gap: 3px;
}

.container {
  /* ... */
  flex: 1 200px;
}
```

This way, the containers won't go below 200px in width, and wrap onto new lines if there is not enough space for them all. This looks neater than the current fixed-width approach.

{{EmbedLiveSample("Reversing the animation direction","100%","250")}}

> Here, the underlying value for all examples' `transform` property is `translateX(50px) rotate(45deg)`. The different `animation-composition` values' effects are as follows:

Reviewer: ```suggestion
The underlying value for the `transform` property in all cases is `translateX(50px) rotate(45deg)`. The different `animation-composition` values' effects are as follows:
```

- With `replace`, the `transform` property in each keyframe entirely replaces the underlying `transform` property set on the animated element. The final effect value for the `transform` property at the `50%` keyframe is `translateY(30px)` (no `rotate` or `translateX`); at the `100%` keyframe, it is `translateX(150px)` (no `rotate` or `translateY`).

>   In essence, the target starts at `transform: translateX(30px) rotate(45deg)` and effectively animates to `transform: translateY(30px)`, then to `transform: translateX(150px)`.

Reviewer: ```suggestion
  The target starts at `transform: translateX(30px) rotate(45deg)` and effectively animates to `transform: translateY(30px)`, then to `transform: translateX(150px)`.
```

> - With `add`, the final effect value at each keyframe would be the original `transform` property with the new one added after.

Reviewer: ```suggestion
- With `add`, the final effect value at each keyframe is the underlying `transform` value with the effect value placed immediately after it.
```

>   Therefore, the target starts at `transform: translateX(30px) rotate(45deg)` and effectively animates to `transform: translateX(30px) rotate(45deg) translateY(30px)` (which is `30px` "downwards" on the rotated Y-axis), then to `transform: translateX(30px) rotate(45deg) translateX(150px)`. Since the additive operation is relative to the underlying `transform` and not the previous keyframe, there is no `translateY(30px)` at `100%`, leaving the element `150px` along the rotated X-axis from the original position.

Reviewer: ```suggestion
  Therefore, the target starts at `transform: translateX(30px) rotate(45deg)` and effectively animates first to `transform: translateX(30px) rotate(45deg) translateY(30px)` (which is `30px` "downwards" on the rotated Y-axis), and then to `transform: translateX(30px) rotate(45deg) translateX(150px)`. Since the additive operation is relative to the underlying `transform` and not the previous keyframe, there is no `translateY(30px)` at `100%`, placing the element `150px` along the rotated X-axis from the original position.
```

- With `accumulate`, the final effect value is the keyframe's effect `transform` combined with the underlying original. At `50%`, `translateY(30px)` combines with the original `translateX(30px)` into a single translation (`translate(30px, 30px)`). At `100%`, the `translateX(150px)` combines with the original `translateX(30px)` to create `translateX(180px)`.

>   Therefore, the target starts at `transform: translateX(30px) rotate(45deg)` and effectively animates to `transform: translate(30px, 30px) rotate(45deg)` then to `transform: translateX(180px) rotate(45deg)`.

Reviewer: ```suggestion
  Therefore, the target starts at `transform: translateX(30px) rotate(45deg)` and effectively animates first to `transform: translate(30px, 30px) rotate(45deg)`, and then to `transform: translateX(180px) rotate(45deg)`.
```

{{EmbedLiveSample("Reversing the animation direction","100%","250")}}

> 

Reviewer: It would be nice to include a lead-up line here, along the lines of:

> The different `animation-composition` value's effects are as follows:

> - With `replace`, the `transform` property in each keyframe will entirely replace that of the original one. The final effect value for the `transform` property at the 50% keyframe would simply be `translateY(30px)` on its own, and at the 100% keyframe would be `translateX(150px)` on its own.

Reviewer: ```suggestion
- With `replace`, the `transform` property in each keyframe entirely replaces the underlying `transform` property set on the animated element. The final effect value for the `transform` property at the `50%` keyframe is `translateY(30px)` (no `rotate` or `translateX`); at the `100%` keyframe, it is `translateX(150px)` (no `rotate` or `translateY`). 
```

I'm not sure if the parentheses are going a bit overboard here, but I wanted to make sure the reader is clear exactly what effects are active at each step.

  Therefore, the target starts at `transform: translateX(30px) rotate(45deg)` and effectively animates to `transform: translateY(30px)`, then to `transform: translateX(150px)`.

- With `add`, the final effect value at each keyframe would be the original `transform` property with the new one added after.

>   Therefore, the target effectively animates to `transform: translateX(30px) rotate(45deg) translateY(30px)` then to `transform: translateX(30px) rotate(45deg) translateX(150px)`.

Reviewer: ```suggestion
  Therefore, the target starts at `transform: translateX(30px) rotate(45deg)` and effectively animates to `transform: translateX(30px) rotate(45deg) translateY(30px)` then to `transform: translateX(30px) rotate(45deg) translateX(150px)`. The `rotate` value is the same in all cases; therefore, the animated element's rotation doesn't change at any point: It stays at `45deg` throughout. Also note that the X and Y translations are relative to the element's rotated X and Y axes, not the page's default X and Y axes.
```

Again, maybe a bit overboard, but I was a bit confused for a minute or so, until I remembered that when rotating an element, the X and Y axes also rotate.

Author: How much would we want this page to go into how composite transforms work? I do think the suggestion specifically is a little much but I wonder if it's this page's responsibility to explain that at all, or if it should be the reader's responsibility to look up the `transform` docs to see how composites work.

I think because the original example tied itself to explaining composite transforms, that's what led to this PR in the first place and could easily have been missed when I was researching the transform problem (as I'm sure I've missed other pages that similarly describe a composite transform process).

Reviewer: I'm not sure if we have a decent explanation of composite transforms anywhere; adding that is definitely beyond the scope of this PR.

For now, how about including brief explanations on this page? I'm happy for you to cut my explanations down a bit if you think they are too much for this page.

It is already much better.

Author: #43496 clarified in the `transform` docs how composite transforms work, including what happens to the coordinate space after each function. That page's [example of composite transform order](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/transform#comparing_the_order_of_transform_functions) now does a much better job of explaining the deal consistently with the rest of the page, and includes an example of rotated axes.

Given the updated `transform` docs, do you feel the `animation-composition` demo should touch on it briefly still?

My gut feeling is that `animation-composition` should focus on describing how the composition works, without explaining how the effect values themselves work, which would be the job of the `transform` docs. i.e. the current explanation assumes you understand how `transform` works and if you don't, the now-improved `transform` docs have you covered.

Reviewer: Oh, OK, great. 

Yeah, this page doesn't need so much explanation in that case. I still think a bit of explanation of what is going on specifically in the example is useful to help follow it.

Author: That's fair, I'll try and expand a little in a streamlined way

> - With `accumulate`, the the final effect value would be the keyframe's `transform` combined with the original. At 50%, `translateY(30px)` combines with the original `translateX(30px)` into a single translation. At 100%, the `translateX(150px)` combines with the original `translateX(30px)`.

Reviewer: ```suggestion
- With `accumulate`, the final effect value is the keyframe's effect `transform` combined with the underlying original. At `50%`, `translateY(30px)` combines with the original `translateX(30px)` into a single translation (`translate(30px, 30px)`). At `100%`, the `translateX(150px)` combines with the original `translateX(30px)` to create `translateX(180px)`.
```

>   Therefore, the target effectively animates to `transform: translate(30px, 30px) rotate(45deg)` then to `transform: translateX(180px) rotate(45deg)`.

Reviewer: ```suggestion
  Therefore, the target starts at `transform: translateX(30px) rotate(45deg)` and effectively animates to `transform: translate(30px, 30px) rotate(45deg)` then to `transform: translateX(180px) rotate(45deg)`. Again, the animated element's `rotation` value stays the same throughout. 
```

Author: I've opted to leave out the bit about rotation since it's not mentioned in the `add` description any more, so seems out of place. And I don't think there's a need to call out the rotation, as discussed before, since that feels more `transform`'s doc's responsibility.

{{EmbedLiveSample("Reversing the animation direction","100%","250")}}

- With `replace`, the final effect value for the `transform` property in the `20%, 40%` keyframe is `translateX(100px)` (completely replacing the underlying value `translateX(30px) rotate(45deg)`). In this case, the element rotates from 45deg to 0deg as it animates from the default value set on the element itself to the non-rotated value set at the 20% mark. This is the default behavior.

> - With `add`, the final effect value for the `transform` property in the `20%, 40%` keyframe is `translateX(30px) rotate(45deg) translateX(100px)`. So the element is first moved 30px to the right, rotated 45deg in place, then moved by 100px along the rotated X axis.

Reviewer: I'm having a bit of trouble understanding what is going on here. Surely, given that the initial `rotate` value is `45deg`, and the animation is animating to `45deg` between 20% and 40%, no rotation happens at all? So I don't know what "rotated 45deg in place" means.

Also, you don't see it being moved to the right by `30px`. Surely the `30px` position change to the right already happens at the beginning, due to the default `transform` value?
